src/services/pin-state-transition/hooks/prepare-activity-log.js

```diff
@@ -78,17 +80,17 @@ const prepareActivityLog = () => (hook) => {
           // So, we will remove `assigned_department` value from the pin.
           action = actions.DENY;
           changedFields.push('assigned_department');
-          previousValues.push(assignedDepartmentObject._id);
-          updatedValues.push(undefined);
+          previousValues.push(assignedDepartmentObject);
+          updatedValues.push(EMAIL_NOTI_NON_ASSIGNED_TEXT);
           description = `${nameOfUser} denies pin #${pinId}`;
         } else if (previousState === states.RESOLVED) {
           // If a pin has already resolved but it is re-opened again.
           // Remove `assigned_department` value from the pin
           // and ORGANIZATION_ADMIN will do assignment again.
           action = actions.RE_OPEN;
           changedFields.push('assigned_department');
-          previousValues.push(assignedDepartmentObject._id);
-          updatedValues.push(undefined);
+          previousValues.push(assignedDepartmentObject);
```

Comment:
You are totally right. This is complicated.
I changed the approach to populate everything inside email-noti function instead.
PTAL @supasate @rapee